### PR TITLE
#49 Fixed false negative for alias directives

### PR DESCRIPTION
**Pull request by eloisetaylor5693**
It seems like static using directives may already be catered for.

If anyone can think of some examples of static directives that give false negatives, then I shall fix.

**Example from `WeakCipherModeAnalyzerTest`**
```
using System;
using System.IO;
using System.Security.Cryptography;
using static System.Security.Cryptography.CipherMode;
using System.Text;

class WeakCipherMode
{
    public static string EncryptECB(string decryptedString)
    {
        DESCryptoServiceProvider desProvider = new DESCryptoServiceProvider();
        desProvider.Mode = ECB;
        desProvider.Padding = PaddingMode.PKCS7;
        desProvider.Key = Encoding.ASCII.GetBytes(""d66cf8"");
        using (MemoryStream stream = new MemoryStream())
        {
            using (CryptoStream cs = new CryptoStream(stream, desProvider.CreateEncryptor(), CryptoStreamMode.Write))
            {
                byte[] data = Encoding.Default.GetBytes(decryptedString);
                cs.Write(data, 0, data.Length);
                return Convert.ToBase64String(stream.ToArray());
            }
        }
    }
}
```

### CODE REVIEW

**Comment by JarLob** [comment]
I wasn't probably clear enough in the description of #49 :(
It started as an optimization because `GetSymbolInfo` is relatively expensive, so checking first if a syntax node name contains the expected text looked as as a good idea. Later I learned that it doesn't work in alias case.
The fix would be to remove
```cs
            var objectCreation = nodeHelper.GetNameNode(ctx.Node);
            if (!objectCreation.ToString().Contains("JavaScriptSerializer"))
                return;
```
and similar code in other analyzers at all and probably to get rid of `GetNameNode` helper too. The helper is used in only two analyzers, so that shouldn't be too hard. But other analyzers should be reviewed for the optimization pattern and unit tests updated ideally for all of them. So it is kind of tedious task :)
Your fix converts whole compilation file to a string and greps it for expected text. Without proper caching it is probably even slower than getting symbolic info from compiler.

**Comment by eloisetaylor5693** [comment]
1/2 way there.  Added unit tests for half the analyzers.  Will get rid of `GetNameNode` in next commits over the next week-or-so

**Comment by JarLob** [comment]
Cool! One suggestion though - you could use `[DataTestMethod]` with `alias` and `name` like:
```cs
[DataTestMethod]
[DataRow("Cookie = System.Web.HttpCookie", "Cookie")]
[DataRow("System.Web",                                    "HttpCookie")]
public async Task CookieWithoutFlags(string alias, string name)
{
    var cSharpTest = $@"
using {alias};
............................................................................................
            var cookie = new {name}(""test"");
```
It doesn't require a creation of new test and makes a good pattern for new tests.

**Comment by JarLob** [comment]
Cool! One suggestion though - you could use `[DataTestMethod]` with `alias` and `name` like:
```cs
[DataTestMethod]
[DataRow("Cookie = System.Web.HttpCookie", "Cookie")]
[DataRow("System.Web",                                    "HttpCookie")]
public async Task CookieWithoutFlags(string alias, string name)
{
    var cSharpTest = $@"
using {alias};
............................................................................................
            var cookie = new {name}(""test"");
```
It doesn't require a creation of new test and makes a good pattern for new tests.

**Comment by JarLob** [comment]
Cool! One suggestion though - you could use `[DataTestMethod]` with `alias` and `name` like:
```cs
[DataTestMethod]
[DataRow("Cookie = System.Web.HttpCookie", "Cookie")]
[DataRow("System.Web",                                    "HttpCookie")]
public async Task CookieWithoutFlags(string alias, string name)
{
    var cSharpTest = $@"
using {alias};
............................................................................................
            var cookie = new {name}(""test"");
```
It doesn't require a creation of new test and makes a good pattern for new tests.

**Comment by JarLob** [comment]
Cool! One suggestion though - you could use `[DataTestMethod]` with `alias` and `name` like:
```cs
[DataTestMethod]
[DataRow("Cookie = System.Web.HttpCookie", "Cookie")]
[DataRow("System.Web",                     "HttpCookie")]
public async Task CookieWithoutFlags(string alias, string name)
{
    var cSharpTest = $@"
using {alias};
............................................................................................
            var cookie = new {name}(""test"");
```
It doesn't require a creation of new test and makes a good pattern for new tests.

File: SecurityCodeScan.Test/Tests/XxeAnalyzerTest.cs [hunk 1]
```diff
@@ -1090,92 +968,18 @@ End Sub
...
-//    Private Shared Settings As New XmlReaderSettings With {{.{property} = {vbValue}}}
-//#Enable Warning BC40000
-
-//    Public Shared Sub parseUpload(path As String)
-//        Dim reader As XmlReader = XmlReader.Create(path, Settings)
-//    End Sub
-//End Class
-//";
-
-//            await VerifyNoWarnings(cSharpTest5, visualBasicTest5).ConfigureAwait(false); // todo: defaults are assumed for member variables
```
[the remark this answers is not in the document]
Comment (eloisetaylor5693) [reply SecurityCodeScan.Test/Tests/XxeAnalyzerTest.cs]: When uncommented, they fail on some scenarios.  I have added the uncommented code back in, and added tests to show the scenarios when false positives are triggered.

**Comment by eloisetaylor5693** [comment]
This is the last function that uses `GetNameNode`.  `GetCallTargetNode` is not used, so shall I just delete it?  Or is there a future need for this logic?

_CSharpSyntaxNodeHelper_
```
        public override SyntaxNode GetCallTargetNode(SyntaxNode node)
        {
            if (node == null)
                return null;

            SyntaxKind kind = node.Kind();
            switch (kind)
            {
                case SyntaxKind.InvocationExpression:
                    ExpressionSyntax callExpr = ((InvocationExpressionSyntax)node).Expression;
                    return GetNameNode(callExpr) ?? callExpr;
                case SyntaxKind.ObjectCreationExpression:
                    return ((ObjectCreationExpressionSyntax)node).Type;
            }

            return null;
        }
```

**Comment by JarLob** [comment]
It is ok to delete it.

**Comment by JarLob** [comment]
Is it done? I see there are no alias cases in `CsrfTokenAnalyzerTest.cs` and OutputCacheAnnotationAnalyzerTest.cs` for example...

**Comment by eloisetaylor5693** [comment]
I just added alias cases where GetNodeName was used, and if appropriate. I can add alias cases for other analysers.

**Comment by JarLob** [comment]
I don't want to torture you :) Thank you!

**Pull request merged**